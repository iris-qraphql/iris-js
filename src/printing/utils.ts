import type { Maybe } from '../utils/type-level';

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
export function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
export const block = (array: ReadonlyArray<string | undefined>) =>
  array.length > 0 ? `{\n${indent(join(array, '\n'))}\n}` : '{}';

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
export function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

export function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}
