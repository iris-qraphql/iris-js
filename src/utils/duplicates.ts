import type { NameNode } from '../language/ast';

export function getDuplicates<T extends { name: NameNode }>(
  items: ReadonlyArray<T>,
): ReadonlyArray<[NameNode, T]> {
  const known: Record<string, NameNode> = {};
  const duplicates: Array<[NameNode, T]> = [];

  for (const item of items) {
    const name = item.name.value;

    if (known[name]) {
      duplicates.push([known[name], item]);
    } else {
      known[name] = item.name;
    }
  }

  return duplicates;
}
