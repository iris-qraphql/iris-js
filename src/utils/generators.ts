import type { GraphQLScalarTypeConfig } from 'graphql';
import { Kind } from 'graphql';

import type { TypeDefinitionNode, TypeNode } from '../types/ast';
import { getVariant } from '../types/ast';
import { IrisKind } from '../types/kinds';
import { buildSchema } from '../types/schema';

export const gqlScalar = <I, O>(
  config: GraphQLScalarTypeConfig<I, O>,
): TypeDefinitionNode => ({
  kind: IrisKind.TYPE_DEFINITION,
  name: { kind: Kind.NAME, value: config.name },
  description: config.description
    ? { kind: Kind.STRING, value: config.description }
    : undefined,
  variants: [],
});

export const sampleTypeRef = (ref: string, defs: string = ''): TypeNode => {
  const { types } = buildSchema(`
    ${defs}
    data MyType = {
      f: ${ref}
    }
  `);

  const type = types.MyType;
  const [field] = getVariant(type)?.fields ?? [];
  return field.type;
};

export const withWrappers = (type: string): Array<string> => [
  type,
  `${type}?`,
  `[${type}]`,
  `[${type}]?`,
  `[${type}?]`,
  `[${type}?]?`,
];
