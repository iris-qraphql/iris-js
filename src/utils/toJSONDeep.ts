import { isObjectLike } from '../jsutils/isObjectLike';
import { mapValue } from '../jsutils/mapValue';

export function toJSONDeep(value: unknown): unknown {
  if (!isObjectLike(value)) {
    return value;
  }

  if (typeof value.toJSON === 'function') {
    return value.toJSON();
  }

  if (Array.isArray(value)) {
    return value.map(toJSONDeep);
  }

  return mapValue(value, toJSONDeep);
}

export function toJSONError(fn: () => unknown) {
  try {
    fn();
    return undefined;
  } catch (error) {
    return toJSONDeep(error);
  }
}
