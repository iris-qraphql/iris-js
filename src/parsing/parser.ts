import { syntaxError } from '../error';
import { Location, Token } from '../types/ast';
import { TokenKind } from '../types/kinds';
import { optional } from '../utils/type-level';

import { isPunctuatorTokenKind, Lexer } from './lexer';
import { Source } from './source';

type FailOptions = {
  expected?: InspectValue;
  unexpected?: Token;
};
export class Parser {
  _lexer: Lexer;

  constructor(source: string) {
    this._lexer = new Lexer(new Source(source));
  }

  node = <T extends { loc?: Location }>(startToken: Token, node: T): T => ({
    ...node,
    loc: new Location(startToken, this._lexer.lastToken, this._lexer.source),
  });

  peek = (kind: TokenKind): boolean => this._lexer.token.kind === kind;

  lookAhead = (): Token => this._lexer.token;

  fail = <T>(options: FailOptions = {}): T => {
    const token = options.unexpected ?? this.lookAhead();
    throw syntaxError(
      this._lexer.source,
      token.start,
      options.expected
        ? `Expected ${inspect(options.expected)}, found ${inspect(token)}.`
        : `Unexpected ${inspect(token)}.`,
    );
  };

  satisfy = (f: (t: Token) => boolean, expected: InspectValue): Token => {
    const token = this.lookAhead();

    if (!f(token)) {
      return this.fail({ expected });
    }

    this._lexer.advance();
    return token;
  };

  token = (kind: TokenKind) =>
    this.satisfy((token) => token.kind === kind, kind);

  keyword = (value: string): void => {
    this.satisfy(
      (token) => token.kind === TokenKind.NAME && token.value === value,
      value,
    );
  };

  manyTill = <T>(f: () => T, closeKind: TokenKind): Array<T> => {
    const nodes = [];
    do {
      nodes.push(f.call(this));
    } while (!optional(this.token, closeKind));
    return nodes;
  };

  many<T>(
    openKind: TokenKind,
    f: () => T,
    closeKind: TokenKind,
  ): ReadonlyArray<T> {
    this.token(openKind);
    return this.manyTill(f, closeKind);
  }

  separatedBy = <T>(delimiter: TokenKind, f: () => T): Array<T> => {
    const nodes = [];
    do {
      nodes.push(f.call(this));
    } while (optional(this.token, delimiter));
    return nodes;
  };
}

type InspectValue = Token | TokenKind | string;

const inspect = (input: InspectValue): string => {
  if (input instanceof Token) {
    return (
      inspect(input.kind) + (input.value != null ? ` "${input.value}"` : '')
    );
  }
  return isPunctuatorTokenKind(input as TokenKind) ? `"${input}"` : input;
};

export type FParser<T> = (parser: Parser) => T;
