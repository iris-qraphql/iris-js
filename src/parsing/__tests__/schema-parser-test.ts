import { dedent } from '../../utils/dedent';
import { toJSONDeep, toJSONError } from '../../utils/toJSONDeep';

import { parse } from '../parser';

function expectSyntaxError(text: string) {
  return expect(toJSONError(() => parse(text)));
}

const snapshot = (doc: string) =>
  expect(toJSONDeep(parse(doc))).toMatchSnapshot();

describe('Schema Parser', () => {
  it('Simple data', () => {
    snapshot(dedent`
      data Hello = {
        world: String
      }
    `);
  });

  it('parses data with description string', () => {
    snapshot(dedent`
      "Description"
      data Hello = {
        world: String
      }
    `);
  });

  it('parses data with description multi-line string', () => {
    snapshot(dedent`
      """
      Description
      """
      # Even with comments between them
      data Hello = {
        world: String
      }
    `);
  });

  it('Description followed by something other than data system definition throws', () => {
    expectSyntaxError('"Description" 1').toEqual({
      locations: [{ column: 15, line: 1 }],
      message: 'Syntax Error: Unexpected Int "1".',
    });
  });

  it('Simple non-null resolver', () => {
    snapshot(dedent`
      data Hello = {
        world: String
      }
    `);
  });

  it('Single value Enum', () => {
    snapshot('data Hello = WORLD {}');
  });

  it('Double value Enum', () => {
    snapshot('data Hello = WO {} | RLD {}');
  });

  it('Simple resolver', () => {
    snapshot('data Hello = World');
  });

  it('Union with two resolvers', () => {
    snapshot('data Hello = Wo | Rld');
  });

  it('Union fails with no resolvers', () => {
    expectSyntaxError('data Hello =  ').toMatchSnapshot();
  });

  it('Union fails with leading pipe', () => {
    expectSyntaxError('data Hello = | Wo | Rld').toMatchSnapshot();
  });

  it('Union fails with double pipe', () => {
    expectSyntaxError('data Hello = Wo || Rld').toMatchSnapshot();
  });

  it('Union fails with trailing pipe', () => {
    expectSyntaxError('data Hello = Wo | Rld |').toMatchSnapshot();
  });

  it('Scalar', () => {
    snapshot('data Hello = String');
  });

  it('Simple data object', () => {
    snapshot(`
    data Hello = {
      world: String
    }`);
  });

  describe('reject reserved names', () => {
    it('rejects an Enum type with incorrectly named values', () => {
      expectSyntaxError(`
        data SomeEnum 
          = __badName {}
        `).toMatchSnapshot();
    });

    it('rejects field arg with invalid names', () => {
      expectSyntaxError(`
        data SomeObject = {
          badField(__badName: String): String
        }
      `).toMatchSnapshot();
    });

    it('rejects an Object type with incorrectly named fields', () => {
      expectSyntaxError(`
        data SomeObject = {
          __badName: String
        }
      `).toMatchSnapshot();
    });
  });
});
