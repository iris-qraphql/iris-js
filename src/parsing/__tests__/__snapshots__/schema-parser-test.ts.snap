// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Parser Double value Enum 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 27,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 10,
          "start": 5,
        },
        "value": "Hello",
      },
      "role": "data",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 18,
            "start": 13,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 15,
              "start": 13,
            },
            "value": "WO",
          },
        },
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 27,
            "start": 21,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 24,
              "start": 21,
            },
            "value": "RLD",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 27,
    "start": 0,
  },
}
`;

exports[`Schema Parser Scalar 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 19,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 10,
          "start": 5,
        },
        "value": "Hello",
      },
      "role": "data",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": undefined,
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 19,
            "start": 13,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 19,
              "start": 13,
            },
            "value": "String",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 19,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple data 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 32,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 10,
          "start": 5,
        },
        "value": "Hello",
      },
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 30,
                "start": 17,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 22,
                  "start": 17,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 30,
                  "start": 24,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 30,
                    "start": 24,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 32,
            "start": 13,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 10,
              "start": 5,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 32,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple data object 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 45,
        "start": 5,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 15,
          "start": 10,
        },
        "value": "Hello",
      },
      "role": "data",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 39,
                "start": 26,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 31,
                  "start": 26,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 39,
                  "start": 33,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 39,
                    "start": 33,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 45,
            "start": 18,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 15,
              "start": 10,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 45,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple field with arg 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 51,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [
                Object {
                  "defaultValue": undefined,
                  "description": undefined,
                  "directives": Array [],
                  "kind": "InputValueDefinition",
                  "loc": Object {
                    "end": 40,
                    "start": 27,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 31,
                      "start": 27,
                    },
                    "value": "flag",
                  },
                  "type": Object {
                    "kind": "NamedType",
                    "loc": Object {
                      "end": 40,
                      "start": 33,
                    },
                    "name": Object {
                      "kind": "Name",
                      "loc": Object {
                        "end": 40,
                        "start": 33,
                      },
                      "value": "Boolean",
                    },
                  },
                },
              ],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 49,
                "start": 21,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 26,
                  "start": 21,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 49,
                  "start": 43,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 49,
                    "start": 43,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 51,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 14,
              "start": 9,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 51,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple field with arg with default value 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 58,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [
                Object {
                  "defaultValue": Object {
                    "kind": "BooleanValue",
                    "loc": Object {
                      "end": 47,
                      "start": 43,
                    },
                    "value": true,
                  },
                  "description": undefined,
                  "directives": Array [],
                  "kind": "InputValueDefinition",
                  "loc": Object {
                    "end": 47,
                    "start": 27,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 31,
                      "start": 27,
                    },
                    "value": "flag",
                  },
                  "type": Object {
                    "kind": "NamedType",
                    "loc": Object {
                      "end": 40,
                      "start": 33,
                    },
                    "name": Object {
                      "kind": "Name",
                      "loc": Object {
                        "end": 40,
                        "start": 33,
                      },
                      "value": "Boolean",
                    },
                  },
                },
              ],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 56,
                "start": 21,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 26,
                  "start": 21,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 56,
                  "start": 50,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 56,
                    "start": 50,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 58,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 14,
              "start": 9,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 58,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple field with list arg 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 54,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [
                Object {
                  "defaultValue": undefined,
                  "description": undefined,
                  "directives": Array [],
                  "kind": "InputValueDefinition",
                  "loc": Object {
                    "end": 43,
                    "start": 27,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 33,
                      "start": 27,
                    },
                    "value": "things",
                  },
                  "type": Object {
                    "kind": "ListType",
                    "loc": Object {
                      "end": 43,
                      "start": 35,
                    },
                    "type": Object {
                      "kind": "NamedType",
                      "loc": Object {
                        "end": 42,
                        "start": 36,
                      },
                      "name": Object {
                        "kind": "Name",
                        "loc": Object {
                          "end": 42,
                          "start": 36,
                        },
                        "value": "String",
                      },
                    },
                  },
                },
              ],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 52,
                "start": 21,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 26,
                  "start": 21,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 52,
                  "start": 46,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 52,
                    "start": 46,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 54,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 14,
              "start": 9,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 54,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple field with two args 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 66,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [
                Object {
                  "defaultValue": undefined,
                  "description": undefined,
                  "directives": Array [],
                  "kind": "InputValueDefinition",
                  "loc": Object {
                    "end": 42,
                    "start": 27,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 33,
                      "start": 27,
                    },
                    "value": "argOne",
                  },
                  "type": Object {
                    "kind": "NamedType",
                    "loc": Object {
                      "end": 42,
                      "start": 35,
                    },
                    "name": Object {
                      "kind": "Name",
                      "loc": Object {
                        "end": 42,
                        "start": 35,
                      },
                      "value": "Boolean",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "description": undefined,
                  "directives": Array [],
                  "kind": "InputValueDefinition",
                  "loc": Object {
                    "end": 55,
                    "start": 44,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 50,
                      "start": 44,
                    },
                    "value": "argTwo",
                  },
                  "type": Object {
                    "kind": "NamedType",
                    "loc": Object {
                      "end": 55,
                      "start": 52,
                    },
                    "name": Object {
                      "kind": "Name",
                      "loc": Object {
                        "end": 55,
                        "start": 52,
                      },
                      "value": "Int",
                    },
                  },
                },
              ],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 64,
                "start": 21,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 26,
                  "start": 21,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 64,
                  "start": 58,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 64,
                    "start": 58,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 66,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 14,
              "start": 9,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 66,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple non-null resolver 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 36,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 34,
                "start": 21,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 26,
                  "start": 21,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 34,
                  "start": 28,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 34,
                    "start": 28,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 36,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 14,
              "start": 9,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 36,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple resolver 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 36,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 34,
                "start": 21,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 26,
                  "start": 21,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 34,
                  "start": 28,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 34,
                    "start": 28,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 36,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 14,
              "start": 9,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 36,
    "start": 0,
  },
}
`;

exports[`Schema Parser Simple resolver 2`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 22,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": undefined,
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 22,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 22,
              "start": 17,
            },
            "value": "World",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 22,
    "start": 0,
  },
}
`;

exports[`Schema Parser Single value Enum 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 21,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 10,
          "start": 5,
        },
        "value": "Hello",
      },
      "role": "data",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 21,
            "start": 13,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 18,
              "start": 13,
            },
            "value": "WORLD",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 21,
    "start": 0,
  },
}
`;

exports[`Schema Parser Union with two resolvers 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 25,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 14,
          "start": 9,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": undefined,
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 19,
            "start": 17,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 19,
              "start": 17,
            },
            "value": "Wo",
          },
        },
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": undefined,
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 25,
            "start": 22,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 25,
              "start": 22,
            },
            "value": "Rld",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 25,
    "start": 0,
  },
}
`;

exports[`Schema Parser parses resolver with description multi-line string 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": Object {
        "block": true,
        "kind": "StringValue",
        "loc": Object {
          "end": 19,
          "start": 0,
        },
        "value": "Description",
      },
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 90,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 68,
          "start": 63,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 88,
                "start": 75,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 80,
                  "start": 75,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 88,
                  "start": 82,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 88,
                    "start": 82,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 90,
            "start": 71,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 68,
              "start": 63,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 90,
    "start": 0,
  },
}
`;

exports[`Schema Parser parses resolver with description string 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": Object {
        "block": false,
        "kind": "StringValue",
        "loc": Object {
          "end": 13,
          "start": 0,
        },
        "value": "Description",
      },
      "directives": Array [],
      "kind": "TypeDefinition",
      "loc": Object {
        "end": 50,
        "start": 0,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 28,
          "start": 23,
        },
        "value": "Hello",
      },
      "role": "resolver",
      "variants": Array [
        Object {
          "description": undefined,
          "directives": Array [],
          "fields": Array [
            Object {
              "arguments": Array [],
              "description": undefined,
              "directives": Array [],
              "kind": "FieldDefinition",
              "loc": Object {
                "end": 48,
                "start": 35,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 40,
                  "start": 35,
                },
                "value": "world",
              },
              "type": Object {
                "kind": "NamedType",
                "loc": Object {
                  "end": 48,
                  "start": 42,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 48,
                    "start": 42,
                  },
                  "value": "String",
                },
              },
            },
          ],
          "kind": "VariantDefinition",
          "loc": Object {
            "end": 50,
            "start": 31,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 28,
              "start": 23,
            },
            "value": "Hello",
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 50,
    "start": 0,
  },
}
`;

exports[`Schema Parser reject reserved names rejects an Enum type with incorrectly named values 1`] = `
Object {
  "locations": Array [
    Object {
      "column": 23,
      "line": 3,
    },
  ],
  "message": "Syntax Error: Name \\"__badName\\" must not begin with \\"__\\", which is reserved by GraphQL introspection.",
}
`;

exports[`Schema Parser reject reserved names rejects an Object type with incorrectly named fields 1`] = `
Object {
  "locations": Array [
    Object {
      "column": 20,
      "line": 3,
    },
  ],
  "message": "Syntax Error: Name \\"__badName\\" must not begin with \\"__\\", which is reserved by GraphQL introspection.",
}
`;

exports[`Schema Parser reject reserved names rejects field arg with invalid names 1`] = `
Object {
  "locations": Array [
    Object {
      "column": 29,
      "line": 3,
    },
  ],
  "message": "Syntax Error: Name \\"__badName\\" must not begin with \\"__\\", which is reserved by GraphQL introspection.",
}
`;
